### PR TITLE
Adds composting with plants

### PR DESCRIPTION
**Pull request by ghost**
You can now compost plant produce by using them on a tray. Composting transfers all reagents from the produce into the tray and deletes the produce.

[qol]

:cl:
 * rscadd: Plant produce can now be used as compost in trays

### CODE REVIEW

**Comment by brndd** [comment]
I don't think it makes sense to compost things in hydroponics trays, which are full of water and therefore aren't a suitable composting environment. Most (all?) places with access to hydroponics trays also already have access to the blender to produce nutrients out of plant matter, or other sources of nutrients.

I think this change would be much better suited for the ghetto botany trays, which are just piles of dirt. That's a much better environment for composting, and since ghetto botany lacks a blender (if I remember correctly), this feature would actually be somewhat useful there.

**Comment by ghost** [comment]
This change applies to both regular trays and soil piles.

The hydroponic trays are clearly advanced machines capable of not being clogged by uranium, plasma, radium, or any other element in existence. Handling a bit of plant matter is inconsequential in comparison, if suspension of disbelief is your issue.

**Comment by brndd** [comment]
All the chemicals put into them from beakers are liquids (through space magic, a regular open beaker is capable of containing even liquid hydrogen at room temperatures and standard pressures), not solids like compost is.

**Comment by ghost** [comment]
Why is space magic good for that but not for this?

**Comment by brndd** [comment]
All good magic needs an internal set of rules and laws it operates on. The spess magic in beakers is that whatever they contain, it's always liquid. It might not be realistic, but it's consistent, and because it's consistent, it makes sense. For the blender (and something like the mortar and pestle which operate pretty much the same), the spess magic is that it always produces liquids. This might not even have to be handwaved by spess magic, as the blending process could just involve throwing away whatever solid waste matter is left.

I could be wrong, but I don't think hydroponics trays have accepted any solid ingredients into them before this. Adding dead plant matter into a hydroponics tray in reality also wouldn't compost it, and it would just form a sediment at the bottom and possibly contaminate the tray. That's why I don't think this change makes sense: there's no pre-established "space magic" rule that says that hydroponics trays accept solids, and it doesn't make sense in terms of realism.

However, for the ghetto botany trays I think this change would be a perfect fit. It's (somewhat) realistic to add compost to them, so it doesn't need any space magic explanations, **and** it's mechanically a good fit because ghetto botanists have fewer options for extracting nutrients out of plants.

**Comment by ghost** [comment]
Whatever man, have it your way. Only soil now supports composting.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
I think doing this with trays is OK but potentially annoying if you accidentally misclick

But anyways: Don't you have to make the user drop the object before deleting it? Or was qdel() updated to deal with that?
Please check what happens if you compost a fruit then try to throw it immediately after

**Comment by codeungabunga** [comment]
I don't see how that is possible, the fruit disappears the moment it is composted and there is nothing to throw. Throwing empty hand doesn't work either. I don't see any runtimes either.

**Comment by ghost** [comment]
```
/obj/item/Destroy()
	if(istype(loc, /mob))
		var/mob/H = loc
		H.drop_from_inventory(src) // items at the very least get unequipped from their mob before being deleted
	if(hasvar(src, "holder"))
		src:holder = null
	for(var/x in actions)
		qdel(x)
	/*  BROKEN, FUCK BYOND
	if(hasvar(src, "my_atom"))
		src:my_atom = null*/
	..()
```

That part is handled in item destroy.

**Comment by I-VAPE-VOX-CLOACA-EVERY-DAY-OF-MY-LIFE** [comment]
Fantastic, good to know.

**Comment by Jubisloviu** [comment]
would be cool if botany had a roundstart compost heap
taking more things out of the biogenerator is great because the biogenerator is a lazy solution to many problems

**Comment by ghost** [comment]
I can code in anything you sprite that does not add more tedium or randomness.

**Comment by Killette2** [comment]
What if you could chuck soil into hydro trays to have the best of both worlds?

**Comment by ghost** [comment]
Trays:
\+ status indicators
\+ tray lid
\+ can be moved
\- cost
\- can't recycle trays

Soil:
\+ cheap
\+ can be recycled at any time (which resets chems etc.)
\- stationary
\- easy to lose plants due to no indicators
\- no lid means strong mutations are extremely risky (plasma from plants)

Both already have their uses and this change is not about differentiating them more, rather giving all plants extra function in form of compost. I plan on adding water to watermelons which will complement this feature, making it possible to gene splice plants that can be used instead of access to fertilizer/water source.

Some food for thought, soil is actually made from `ore/glass` bricks, which is essentially sand that can be grown from grass. The entire previous sentence is complete magic.

File: code/modules/hydroponics/hydro_tray.dm [hunk 1]
```diff
@@ -387,6 +387,11 @@
 	else if((O.sharpness_flags & (SHARP_BLADE|SERRATED_BLADE)) && harvest)
 		attack_hand(user)
 
+	else if(istype(O, /obj/item/weapon/reagent_containers/food/snacks/grown)) //composting
+		to_chat(user, "You use [O] as compost for [src].")
```
Comment (MadmanMartian) [inline code/modules/hydroponics/hydro_tray.dm]: "you use \the [O] as compost for \the [src]."

Comment (ghost) [reply code/modules/hydroponics/hydro_tray.dm]: Done.

**Review by MadmanMartian — APPROVED** [review]

**Pull request merged**